### PR TITLE
DAV endpoint

### PR DESCRIPTION
**Pull request by PVince81**
- [x] REQUIRES: https://github.com/owncloud/customgroups/pull/1 to be merged

This PR is based on top of https://github.com/owncloud/customgroups/pull/1.

This adds DAV endpoints for https://github.com/owncloud/customgroups/issues/3

### CODE REVIEW

**Comment by PVince81** [comment]
Rebased, added lots of tests.

A few calls and conditions are missing still.

**Comment by PVince81** [comment]
Already implemented the main part of the spec https://github.com/owncloud/customgroups/issues/3 for managing groups and memberships.

Next step: REPORT

**Comment by PVince81** [comment]
I've added the REPORT as well:

Here is how to get one.
Put this into `report-customgroups.xml`:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<oc:filter-members xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns" >
	<oc:filter-rules>
		<oc:user-id>user1</oc:user-id>
		<!--	<oc:is-admin>1</oc:is-admin> -->
	</oc:filter-rules>
</oc:filter-members>
```
Then `curl -u admin:admin -X REPORT -H "Content-Type: text/xml" --data-binary "@report-customgroups.xml" 'http://localhost/owncloud/remote.php/dav/customgroups/'`

Gives:
```xml
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:cal="urn:ietf:params:xml:ns:caldav" xmlns:cs="http://calendarserver.org/ns/" xmlns:card="urn:ietf:params:xml:ns:carddav" xmlns:oc="http://owncloud.org/ns">
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/dav/customgroups/x/user1</d:href>
  <d:propstat>
   <d:prop>
    <oc:is-admin>0</oc:is-admin>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:status>HTTP/1.1 200 OK</d:status>
  <d:href>/owncloud/remote.php/dav/customgroups/y/user1</d:href>
  <d:propstat>
   <d:prop>
    <oc:is-admin>0</oc:is-admin>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>
```

It basically returns the result for the `CustomGroupMemberNode`. I'm not too happy about this because it will not be useful for the UI. It misses the group display name and the group URI. Well, the group URI can be extracted from the result path.

I'd like to refine this separately once we start working on the UI.

**Comment by PVince81** [comment]
- [ ] make it possible for regular admins to do the same things like group admins

The sad thing about adding this is that now I need to inject a group manager everywhere...

File: lib/Dav/CustomGroupsPlugin.php [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+	/**
+	 * This initializes the plugin.
+	 *
+	 * This function is called by Sabre\DAV\Server, after
+	 * addPlugin is called.
+	 *
+	 * This method should set up the required event subscriptions.
+	 *
+	 * @param Server $server
```
Comment (DeepDiver1975) [inline lib/Dav/CustomGroupsPlugin.php]: Server is unknown

File: lib/Dav/CustomGroupsPlugin.php [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	}
+
+	/**
+	 * REPORT operations to look for comments
+	 *
+	 * @param string $reportName
+	 * @param [] $report
+	 * @param string $uri
+	 * @return bool
+	 * @throws NotFound
```
Comment (DeepDiver1975) [inline lib/Dav/CustomGroupsPlugin.php]: NotFound is unknown

File: lib/Dav/CustomGroupsPlugin.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	 * @throws ReportNotSupported
```
Comment (DeepDiver1975) [inline lib/Dav/CustomGroupsPlugin.php]: ReportNotSupported is unknown

File: lib/Dav/CustomGroupsPlugin.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	 */
+	public function onReport($reportName, $report, $uri) {
+		$node = $this->server->tree->getNodeForPath($uri);
+		if(!$node instanceof RootCollection || $reportName !== self::REPORT_NAME) {
+			throw new ReportNotSupported();
```
Comment (DeepDiver1975) [inline lib/Dav/CustomGroupsPlugin.php]: ReportNotSupported is unknown

File: lib/Dav/RootCollection.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
+<?php
+/**
+ * @author Vincent Petry <pvince81@owncloud.com>
+ *
+ * @copyright Copyright (c) 2016, ownCloud GmbH.
```
Comment (DeepDiver1975) [inline lib/Dav/RootCollection.php]: remove the .

File: lib/Dav/RootCollection.php [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	 * @var IUserManager
+	 */
+	private $userManager;
+
+	/**
+	 * @var IUserSession
+	 */
+	private $userSession;
+
+	/**
```
Comment (DeepDiver1975) [inline lib/Dav/RootCollection.php]: PHPDoc missing

File: lib/Dav/CustomGroupMemberNode.php [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+				self::PROPERTY_IS_ADMIN => $this->memberInfo['is_admin'] ? 1 : 0,
+			];
+		}
+		return [];
+	}
+
+	/**
+	 * Update the admin flag
+	 *
+	 * @param int $isAdminPropValue 1 for true, 0 for false
```
Comment (DeepDiver1975) [inline lib/Dav/CustomGroupMemberNode.php]: return tag is missing

**Comment by DeepDiver1975** [comment]
> It basically returns the result for the CustomGroupMemberNode. I'm not too happy about this because it will not be useful for the UI. It misses the group display name and the group URI. Well, the group URI can be extracted from the result path.

please have a look at the carddav query request - there is a node where you select the elements which you expect to be returned ...

https://tools.ietf.org/html/rfc6352#section-8.6.3

```xml
     <D:prop>
       <D:getetag/>
       <CG:display/>
     </D:prop>

**Comment by PVince81** [comment]
Squashed and removed the composer files I checked in by mistake from the other branch.

**Comment by PVince81** [comment]
ownCloud admins can now also manage groups even when they aren't members.

Instead of passing around yet another object `IGroupManaber`, I created a helper class instead `MembershipHelper` which takes care of checking whether the current user can or cannot manage stuff based on their role and memberships.

**Comment by PVince81** [comment]
@DeepDiver1975 please re-review.

My next step is to adjust the REPORT to implement pagination, but this could also be done separately if this PR is already mergable.

**Comment by PVince81** [comment]
I'm feeling a bit lazy with the REPORT + pagination especially if I need to do it for all the different endpoints. So I came up with this generic paginated report idea, please check it out: https://github.com/owncloud/core/issues/26840

**Comment by PVince81** [comment]
REPORT will be done in a separate PR.

@DeepDiver1975 please review

**Comment by DeepDiver1975** [comment]
- [ ] listing users does not work /remote.php/dav/customgroups/users/

**Comment by PVince81** [comment]
> listing users does not work /remote.php/dav/customgroups/users/

This is by design, same like the "users" list principal.

You need to know the user name you want to query.

**Comment by DeepDiver1975** [comment]
> This is by design, same like the "users" list principal.

Okay - in debug mode we allow listing of users in other collections as well

**Comment by PVince81** [comment]
> Okay - in debug mode we allow listing of users in other collections as well

Yet another code path to test / unit test and maintain. I'd rather not.

Unless we make it more useful by enabling user listing for any admins
And even then, the admin should be able to get a listing from the other endpoint.

Let me know what you think.

**Comment by PVince81** [comment]
Adjusted exception formatting.

Added missing case: ownCloud admin can now enter the "users/$userId" subcollection even if not a member.

@DeepDiver1975 I'd like to not add user listing here, it would require injecting the user manager, etc. I don't think it's useful enough, especially if only used in debug mode.

**Pull request merged**